torch/amp/grad_scaling.py

```diff
@@ -0,0 +1,382 @@
+                     scale_growth_factor=_default_scale_growth_factor,
+                     scale_backoff_factor=_default_scale_backoff_factor,
+                     scale_scheduler=None,
+                     skip_if_inf=True,
+                     found_inf=None,
+                     scaling_enabled=True):
+    r"""
+    :func:`unscale_and_step` is patched onto an optimizer instance by :func:`add_amp_attributes`.  Since
+    :func:`unscale_and_step` becomes a method, ``self`` is passed automatically. :func:`unscale_and_step` should only
+    be called as an attribute of its optimizer instance (e.g. ``opt.unscale_and_step()``), not as a free function.
+
+    :func:`unscale_and_step` unscales the gradients owned by this optimizer (dividing them by ``current_scale``) and
+    checks them for infs/nans.  By default, if no infs/nans are found, it runs the underlying ``self.step()``.
+    If infs/nans are found, it skips the step().
+
+    Based on whether infs/nans were found, :func:`unscale_and_step` then uses the current scale to compute a new
+    recommended scale for next iteration.  By default, the scale adjustment is the following::
+
+       if gradients have inf/nan:
+           recommended scale  = current_scale*scale_backoff_factor
+       else:
+           recommended_scale = current_scale*scale_growth_factor
+
+    Default values for ``scale_backoff_factor`` and ``scale_growth_factor`` are 0.5 and 1.001 respectively.
+    This means that if no infs/nans were found, the new recommended scale is slightly increased from the current scale,
+    and if infs/nans were found, the new recommended scale is half the current scale.
+
+    To minimize CPU-GPU syncs, which can impair GPU utilization, :func:`unscale` returns ``found_inf`` and
+    ``recommended_scale`` as device Tensors rather than Python floats.  If you wish to inspect whether an
+    inf/nan was found, or the recommended next scale value, query the ``.item()`` attributes of the returned Tensors.
+    In general this should be done sparingly (not every iteration), because ``.item()`` incurs a CPU-GPU sync.
+
+    .. warning::
+            :func:`unscale_and_step` should only be called after all gradients for this optimizer's ``step`` have been
+            populated.
+
+    Arguments:
+        current_scale (float or torch.cuda.FloatTensor):  The value most recently used to scale gradients
+            (by e.g. a call to :func:`amp.scale_outputs`).
+        scale_growth_factor (float, optional, default=1.001):  A Python float to be used as the scale growth factor.
+            If ``scale_scheduler`` is supplied, this argument will be ignored.
+        scale_backoff_factor (float, optional, default=0.5):  A Python float to be used as the scale backoff factor.
+            If ``scale_scheduler`` is supplied, this argument will be ignored.
+        scale_scheduler (callable, optional):  Function to customize how the recommended scale is updated.
```

Comment:
Are there examples of networks using a scale_scheduler?

Reply:
No, I thought it would be a nice convenience feature.  

However, since the scale factor is fully visible and controllable by the user, I don't think it's any easier for people to define and supply a scale scheduler than it would be to just manipulate the scale directly.  I guess this is just API clutter and can be gotten rid of.

Reply:
Or added later.